src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,273 @@
+{
+    internal sealed class TranscodingReadStream : Stream
+    {
+        private static readonly int OverflowBufferSize = Encoding.UTF8.GetMaxByteCount(1); // The most number of bytes used to represent a single UTF char
+
+        internal const int MaxByteBufferSize = 4096;
+        internal const int MaxCharBufferSize = 3 * MaxByteBufferSize;
+
+        private readonly Stream _stream;
+        private readonly Decoder _decoder;
+
+#if NETSTANDARD2_0
+        private readonly Encoder _encoder;
+#endif
+
+        private ArraySegment<byte> _byteBuffer;
+        private ArraySegment<char> _charBuffer;
+        private ArraySegment<byte> _overflowBuffer;
+        private bool _disposed;
+
+        public TranscodingReadStream(Stream input, Encoding sourceEncoding)
+        {
+            _stream = input;
+
+            // The "count" in the buffer is the size of any content from a previous read.
+            // Initialize them to 0 since nothing has been read so far.
+            _byteBuffer = new ArraySegment<byte>(
+                ArrayPool<byte>.Shared.Rent(MaxByteBufferSize),
+                0,
+                count: 0);
+
+            // Attempt to allocate a char buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the byte -> char conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            // The conversion process is tolerant of char buffer that is not large enough to convert all the bytes at once.
+            int maxCharBufferSize = Math.Min(MaxCharBufferSize, sourceEncoding.GetMaxCharCount(MaxByteBufferSize));
+            _charBuffer = new ArraySegment<char>(
+                ArrayPool<char>.Shared.Rent(maxCharBufferSize),
+                0,
+                count: 0);
+
+            _overflowBuffer = new ArraySegment<byte>(
+                ArrayPool<byte>.Shared.Rent(OverflowBufferSize),
+                0,
+                count: 0);
+
+            _decoder = sourceEncoding.GetDecoder();
+
+#if NETSTANDARD2_0
+            _encoder = Encoding.UTF8.GetEncoder();
+#endif
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        internal int ByteBufferCount => _byteBuffer.Count;
+        internal int CharBufferCount => _charBuffer.Count;
+        internal int OverflowCount => _overflowBuffer.Count;
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
```

Comment:
Why is this not supported?  This should be made to work.

Reply:
(We should also override the other reading methods, like ReadByte, Read(Span), ReadAsync(Memory), Begin/EndRead, etc.)

Reply:
Why do we need to implement them? This is an internal type and is implementation detail to only use the async methods (`ReadAsync`). Why do we need to implement something that we are not going to use.

Reply:
If it's only used internally by JsonSerializer.DeserializeAsync, then it's fine.

Note that HttpClient is getting sync methods, though, so the sync methods on the write stream will be usable.

Reply:
Can we do that on a follow-up PR that applies all the necessary changes that are only valid for .NET 5?